stablehlo/reference/Element.cpp

```diff
@@ -212,6 +223,76 @@ Element floor(const Element &e) {
   return Element(e.getType(), FloatAttr::get(e.getType(), val));
 }
 
+Element max(const Element &e1, const Element &e2) {
+  return map(
+      e1, e2,
+      [](APInt lhs, APInt rhs) { return llvm::APIntOps::smax(lhs, rhs); },
+      [](APInt lhs, APInt rhs) { return llvm::APIntOps::umax(lhs, rhs); },
+      [](APFloat lhs, APFloat rhs) { return llvm::maximum(lhs, rhs); },
+      [](std::complex<APFloat> lhs, std::complex<APFloat> rhs) {
+        auto isFloatGreaterThan = [](APFloat A,
+                                     APFloat B) -> llvm::Optional<bool> {
+          // APFloat::compare treats -0.0 == 0.0, but IEEE 754-2018 demands -0 <
+          // +0.
+          if (A.isZero() && B.isZero() && (A.isNegative() != B.isNegative()))
+            return !A.isNegative();
+
+          auto cmpRes = A.compare(B);
+          if (cmpRes == APFloat::cmpGreaterThan) return true;
+          if (cmpRes == APFloat::cmpLessThan) return false;
+          if (cmpRes == APFloat::cmpUnordered) return A.isNaN();
+          return llvm::NoneType();
+        };
+
+        // Lexicographic comparison starts with real parts.
+        auto cmpReal = isFloatGreaterThan(lhs.real(), rhs.real());
+        if (cmpReal) return *cmpReal ? lhs : rhs;
+
+        // Real parts are equal, compare the imag parts.
+        auto cmpImag = isFloatGreaterThan(lhs.imag(), rhs.imag());
+        if (cmpImag) return *cmpImag ? lhs : rhs;
+
+        // Both real and imag parts are equal, return any (say the left
+        // operand).
+        return lhs;
+      });
+}
+
+Element min(const Element &e1, const Element &e2) {
+  return map(
+      e1, e2,
+      [](APInt lhs, APInt rhs) { return llvm::APIntOps::smin(lhs, rhs); },
+      [](APInt lhs, APInt rhs) { return llvm::APIntOps::umin(lhs, rhs); },
```

Comment:
Let's use a conditional instead of introducing another lambda parameter in `map`. This will avoid duplication in other ops which don't need to branch on signedness. The conditional could look along the lines of:

```
[&](APInt lhs, APInt rhs) {
  e1.getType().isSignedInteger() ? llvm::APIntOps::smin(lhs, rhs) : llvm::APIntOps::umin(lhs, rhs);
}
```